service/worker/replicator/historyRereplicator.go

```diff
@@ -0,0 +1,438 @@
+)
+
+type (
+	// HistoryRereplicator is the interface for resending history events to remote
+	HistoryRereplicator interface {
+		// SendMultiWorkflowHistory sends multiple run IDs's history events to remote
+		SendMultiWorkflowHistory(domainID string, workflowID string,
+			beginingRunID string, beginingFirstEventID int64, endingRunID string, endingNextEventID int64) error
+	}
+
+	// HistoryRereplicatorImpl is the implementation of HistoryRereplicator
+	HistoryRereplicatorImpl struct {
+		domainCache    cache.DomainCache
+		frontendClient frontend.Client
+		historyClient  h.Client
+		serializer     persistence.HistorySerializer
+		logger         bark.Logger
+	}
+)
+
+// NewHistoryRereplicator create a new HistoryRereplicatorImpl
+func NewHistoryRereplicator(domainCache cache.DomainCache, frontendClient frontend.Client, historyClient h.Client,
+	serializer persistence.HistorySerializer, logger bark.Logger) *HistoryRereplicatorImpl {
+
+	return &HistoryRereplicatorImpl{
+		domainCache:    domainCache,
+		frontendClient: frontendClient,
+		historyClient:  historyClient,
+		serializer:     serializer,
+		logger:         logger,
+	}
+}
+
+// SendMultiWorkflowHistory sends multiple run IDs's history events to remote
+func (h *HistoryRereplicatorImpl) SendMultiWorkflowHistory(domainID string, workflowID string,
+	beginingRunID string, beginingFirstEventID int64, endingRunID string, endingNextEventID int64) (err error) {
+	// NOTE: begining run ID and ending run ID can be different
+	// the logic will try to grab events from begining run ID, first event ID to ending run ID, ending next event ID
+
+	// beginingRunID can be empty, if there is no workflow in DB
+	// endingRunID must not be empty, since this function is trigger messing events of endingRunID
+
+	runID := beginingRunID
+	for len(runID) != 0 && runID != endingRunID {
+		firstEventID, nextEventID := h.eventIDRange(runID, beginingRunID, beginingFirstEventID, endingRunID, endingNextEventID)
+		runID, err = h.sendSingleWorkflowHistory(domainID, workflowID, runID, firstEventID, nextEventID)
+		if err != nil {
+			return err
+		}
+	}
+
+	if runID == endingRunID {
+		firstEventID, nextEventID := h.eventIDRange(runID, beginingRunID, beginingFirstEventID, endingRunID, endingNextEventID)
+		_, err = h.sendSingleWorkflowHistory(domainID, workflowID, runID, firstEventID, nextEventID)
+		return err
+	}
+
+	// we have runID being empty string
+	// this means that beginingRunID to endingRunID does not have a continue as new relation ship
+	// such as beginingRunID -> runID1, then runID2 (no continued as new), then runID3 -> endingRunID
+	// need to work backwords using endingRunID,
+	// moreover, in the above case, runID2 cannot be replicated since no one points to it
+
+	// runIDs to be use to resend history, in reverse order
+	runIDs := []string{endingRunID}
+	runID = endingRunID
+	for len(runID) != 0 {
+		runID, err = h.getPrevRunID(domainID, workflowID, runID)
```

Comment:
need to handle the edge case where the initial runID is already deleted from active cluster.

Reply:
if the beginning run ID is already deleted from remote, then there is nothing this logic can do, since current cluster needs those events. manual operation must be done.

the only edge case this logic can handle is when tracing back using the ending run ID, and encounter EntityNotExistsError. this means that a workflow is not replicated to the currently cluster and the same workflow is also missing in remote.